### ISSUE TITLE
new Event API for internal data hash mutation

### ISSUE THREAD

**Issue by colinsurprenant** [elastic/logstash#4326]
This issue reboots issue #4293 and relates to #4191 

The goal is to add `Event` methods for the plugins to correctly manipulate the underlying data hash by abstracting away the internal implementation details and moving away from mutable interactions with the Event#to_hash returned object.

The current `Event#append` and `Event#overwrite` takes an `Event` as argument. This is fine for some plugins and AFAIK `Event#append` is only used by multiline. We can have a separate discussion about these. 

I suggest adding `from_hash` and `merge_hash`:

- `Event#from_hash` would replace  the `Event` data completely:
```ruby
event # contains { :message => "hello world" }
hash # contains { :foo => "bar" }
event.from_hash(hash) # contains { :foo => "bar" }  }
```

- `Event#merge_hash` would merge in external hash content:
```ruby
event # contains { :message => "hello world" }
hash # contains { :foo => "bar" }
event.merge_hash(hash) # contains { :message => "hello world", :foo => "bar" }
```

Once this is in place, the following plugins will need followup to use the new API
- [ ] filter-kv, see https://github.com/logstash-plugins/logstash-filter-kv/pull/21
- [ ] filter-geoip, see https://github.com/logstash-plugins/logstash-filter-geoip/pull/56
- [ ] filter-json, see https://github.com/logstash-plugins/logstash-filter-json/pull/18
- [ ] filter-csv, https://github.com/logstash-plugins/logstash-filter-csv/pull/20
- [ ] filter-useragent, see https://github.com/logstash-plugins/logstash-filter-useragent/pull/20

**Comment by colinsurprenant** [comment]
I would like to bump this as we are closing in 5.0 and would like to see if there is any traction to this proposal.

Rationale recap:

#### `Event#merge_hash`

The Ruby signature would be 
```ruby
def merge_hash(data, target_fieldref = nil)
```

Where `data` is a Hash and `target_fieldref` would be the target field where to merge the data hash and if omitted would merge at the Event root. 

Many plugins, mainly filters, produces a nested data structures, **think geoip filter**, which need to be merged into the event. Currently setting nested values can be done using:

1. going through individual values and setting each value using its field reference.

 ```ruby
 event.set("[foo][bar]", 1)
 event.set("[foo][baz]", 2)
 ```

2. building a result hash and **overwriting** that result value at a *destination* field reference. 

 ```ruby
 result = {}
 result["bar"] = 1
 result["baz"] = 2
 event.set("foo", result)
 ```

3. building a result hash **merging** that result value at a *destination* field reference. 

 ```ruby
 result = {}
 result["bar"] = 1
 result["baz"] = 2
 current = event.get("foo") || {}
 event.set("foo", current.merge(result))
 ```
 - what do we do if the current value at `[foo]` is a scalar? developer will have to add specific checks if this is a potential problem.

- The proposal to introduce the `merge_hash` would help case 3.

 ```ruby
 result = {}
 result["bar"] = 1
 result["baz"] = 2
 event.merge_hash(result, "foo")
 ```
 - what do we do if the current value at `[foo]` is a scalar? behaviour will have to be decided in `merge_hash`.

#### `Event#from_hash`

The Ruby signature would be 
```ruby
def from_hash(data)
```

Where `data` is a Hash. The Event inner data would be replaced by the passed `data`. 

Again, this is mainly in the context of filters to provide the ability to basically swap-in an event content without having to **cancel** the current event and **yeild** a new event. **Think json filter**.

I am not sure how we should deal with the `@metadata`, `message`, `type` and `tags` fields if they are not part of the supplied data hash, should we preserve them or really completely wipe the Event inner data?

**Comment by jordansissel** [comment]
re: `from_hash` - I think this is not an API we need given you filters can do cancel+yield. With the json filter example, the current behavior with no `@target` is to [merge the JSON hash](https://github.com/logstash-plugins/logstash-filter-json/blob/master/lib/logstash/filters/json.rb#L101) to the root of the event, not replacing it -- I mean, it uses a `merge_hash`-style approach.

I agree there may be use cases for a "replace the event data" but I'm not yet convinced we need a special API (`from_hash`) with it. We had (have?) a way for [one event to overwrite another](https://github.com/elastic/logstash/blob/v1.5.0/lib/logstash/event.rb#L167-L177) but that method doesn't give me good feelings either.

**Comment by ph** [comment]
> what do we do if the current value at [foo] is a scalar? behaviour will have to be decided in merge_hash.

Shouldnt we raise an incompatible type for the merge?

`Event#from_hash`, I think the behavior is a bit confusing if we keep some date of the original object.
Its probably clearer to do the following.

We want a new event: just instantiate a new one?
You want to keep some part of the original event, update it with the new data.?

**Comment by colinsurprenant** [comment]
yes, I agree with the `from_hash` comments ^^. I am totally not sure about the usefulness.
I just checked and only multiline filter and zeromq filter uses the `Event#overwrite` method @jordansissel  - we should probably get rid of it at some point.

**Comment by jordansissel** [comment]
@colinsurprenant ++ on removing Event#overwrite (possibly in time for Logstash 5.0?)
